Synchronize view limits of shared axes after setting ticks

l-johnston opened the pull request:
## PR Summary
Closes #8946

Synchronization uses the approach implemented in PR #18308 and tests according to the one outlined in issue #8946.
## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [N/A] New features are documented, with examples if plot related.
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and `pydocstyle<4` and run `flake8 --docstring-convention=all`).
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

jklymak commented on lib/matplotlib/axis.py:
```diff
@@ -1831,21 +1831,38 @@ def set_ticks(self, ticks, minor=False):
+        else:
+            shared = [self]
+        for axis in shared:
+            if len(ticks) > 1:
```
It seems that maybe `xleft` and `xright` should be set outside the loop in case `xleft` and `xright` have been swapped on a shared axes somehow?  I assume the idea is for all the axes to be identical.

jklymak reviewed (commented):
I'm a little leery of this behaviour....  You can imagine a user wanting exactly the opposite: they have some shared axes, but they want one of the axes to have different explicit ticks, but keep the same x/y limits.  I didn't read the original issue - is this really a change we want?

jklymak commented:
@l-johnston Thanks again for your work on this.  On the dev call @anntzer and @efiring pointed out that a) this is the correct functionality, so please forgive my concern, b) this PR could likely be made a *lot* simpler by just calling `self.set_xlim` or `self.set_ylim`, which have all the axis view logic in them already.  ie. no need to do the loops in this PR.

QuLogic commented:
I'm not sure it'd be so simple, as this is in the `Axis` class, not `XAxis` or `YAxis`, so it doesn't  know which one to call. Even the implementation here is not quite correct, as it doesn't handle z axis sharing on 3D plots.

anntzer commented:
> so it doesn't know which one to call.

You can do something along the lines of
```python
set_xlim(...) if self is self.axes.xaxis else
set_ylim(...) if self is self.axes.yaxis else
set_zlim(...) if self is getattr(self.axes, "zaxis") else
(fall back on the old way of using set_view_interval)
```
or
```python
try:
    axis_name, = [k for k, v in self.axes._get_axis_map() if v is self]
except ValueError:
    (fall back)
getattr(self, f"set_{axis_name}lim")(...)
```

btw instead of set_x/y/zlim you can also use set_x/y/zbound which will auto-handle inverted axises for you.

jklymak commented:
Sorry @l-johnston I think your rebase got messed up.

l-johnston commented:
@jklymak I implemented @anntzer 's approach of using set_x/y/zbound, but this breaks 20 tests such as:
_______________________________________________ test_set_ticks_inverted ________________________________________________
```python
    def test_set_ticks_inverted():
        fig, ax = plt.subplots()
        ax.invert_xaxis()
        ax.set_xticks([.3, .7])
>       assert ax.get_xlim() == (1, 0)
E       assert (0.7, 0.3) == (1, 0)
E         At index 0 diff: 0.7 != 1
E         Use -v to get the full diff

lib/matplotlib/tests/test_axes.py:6520: AssertionError
```

anntzer commented:
Well, perhaps my suggestion was wrong, I didn't actually try it :)

jklymak commented:
Well, I'm confused - does setting the xticks set the limits to the first and last tick or does it insert a margin?

l-johnston commented:
@jklymak setting ticks inserts some margin since the call to `set_view_interval` has default `ignore=False` and this behavior is described in the set_view_interval docstring. I can cause the exact same failures by setting `ignore=True`. So, to prevent breaking existing tests, I think we have to stay with the set_view_interval approach. I have updated it to accommodate @QuLogic 's 3D case and added a test.
